src/test/java/test/dataprovider/DataProviderRetryAnalyzer.java

```diff
@@ -0,0 +1,19 @@
+package test.dataprovider;
+
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+public class DataProviderRetryAnalyzer implements IRetryAnalyzer {
+    private int currentTry = 0;
+    private int maxreruntimes = 1;
+    @Override
+    public synchronized boolean retry(ITestResult result) {
```

Comment:
Maybe we can use `AtomicInteger` and get rid of `synchronized` keyword ? On a side note, I noticed that you haven't enabled the `parallel` attribute on `@DataProvider` annotation. So you can completely get rid of the need for ensuring Thread safety.